### PR TITLE
[RSDK-5863]   Add Address Sanitizer (ASan) option for intel realsense module

### PR DESCRIPTION
**Pull request by hexbabe**
[RSDK-5863](https://viam.atlassian.net/browse/RSDK-5863)

Originally, we wanted to add CI for ASAN, but since CI runners don't come with Intel Realsenses + it is difficult to software emulate the Realsense + adding mocks & unit tests would be quite unwieldy and difficult, we are just gonna support ASAN/LSAN as a build option.

This PR enables developers to run something like `canon -arch arm64 make clean appimage-arm64 SANITIZE=ON` to build the module with address sanitizer on. This allows ASAN/LSAN to analyze the code as it runs with viam-server.

## Testing

On linux/arm64

Commented out `tjDestroy(handle);` in `camera_realsense.cpp` `encodeJPEG` function to intentionally introduce a similar leak fixed in https://viam.atlassian.net/browse/RSDK-5869.

Ran `canon -arch arm64 make clean appimage-arm64 SANITIZE=ON`

```
4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ #3 0xaaaade725e94 in viam::realsense::CameraRealSense::get_image(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<viam::sdk::ProtoType>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<viam::sdk::ProtoType> > > > > const&) (/tmp/.mount_viam-c7oFcxj/usr/bin/viam-camera-realsense+0x105e94)

4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ #2 0xaaaade712dc0 in (anonymous namespace)::encodeJPEGToResponse(unsigned char const*, unsigned int, unsigned int) (/tmp/.mount_viam-c7oFcxj/usr/bin/viam-camera-realsense+0xf2dc0)

4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ #1 0xffffa7f9ae90 in tjInitCompress (/tmp/.mount_viam-c7oFcxj/usr/lib/aarch64-linux-gnu/libturbojpeg.so.0+0x3ae90)

4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ #0 0xaaaade6cac90 in __interceptor_calloc (/tmp/.mount_viam-c7oFcxj/usr/bin/viam-camera-realsense+0xaac90)

4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ Direct leak of 939456 byte(s) in 504 object(s) allocated from:

4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_
4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ ==110799==ERROR: LeakSanitizer: detected memory leaks

4/26/2024, 2:12:53 PM error robot_server.modmanager.process.irs-local-mod_/home/rawr-ubuntu/Projects/viam-camera-realsense/packaging/appimages/deploy/viam-camera-realsense-latest-aarch64.AppImage.StdErr pexec/managed_process.go:252 \_ =================================================================
```

[RSDK-5863]: https://viam.atlassian.net/browse/RSDK-5863?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: integration/tests/all_test.go [hunk 1]
```diff
@@ -28,6 +28,14 @@ func TestCameraServer(t *testing.T) {
 		_, _, err = cam.Images(context.Background())
 		test.That(t, err, test.ShouldBeNil)
 	})
+	t.Run("get image method", func(t *testing.T) {
```
Comment (hexbabe) [inline integration/tests/all_test.go:31]: slightly off topic-- but more test coverage is good

File: packaging/appimages/realsense-integration-tests-aarch64.yml [hunk 2]
```diff
@@ -29,6 +29,7 @@ AppDir:
     - sourceline: deb [trusted=yes] http://deb.debian.org/debian bookworm-updates main
     include:
     - libturbojpeg0:arm64
+    - libjpeg62-turbo:arm64
```
Comment (hexbabe) [inline packaging/appimages/realsense-integration-tests-aarch64.yml:32]: this was missing for some reason. running the binary on arm64 fails without it

**Review by randhid — APPROVED** [review]
Adding Bijan/nick s because I can't give it a thorough read through,, but lgtm!

**Comment by randhid** [comment]
For context @bhaney @nicksanford the criterea for this is to ensure that our devs don't introduce memory leaks, we're not going to battle test real sense's code for them and address issues on their side if there is ever a customer blocking bug.

File: CMakeLists.txt [hunk 3]
```diff
@@ -32,6 +35,18 @@ endif()
...
+if(ENABLE_SANITIZER)
+    add_compile_options(-fsanitize=address)
+    add_link_options(-fsanitize=address)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -static-libasan")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -static-libasan")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static-libasan")
+    target_compile_options(viamrealsense PRIVATE -fsanitize=address)
+    target_link_options(viamrealsense PRIVATE -fsanitize=address)
+    target_compile_options(viam-camera-realsense PRIVATE -fsanitize=address)
+    target_link_options(viam-camera-realsense PRIVATE -fsanitize=address)
```
Comment (nicksanford) [inline CMakeLists.txt:47]: How do you know these are the correct options to enable ASan? Is there a guide you are working from?

File: CMakeLists.txt [hunk 4]
```diff
@@ -32,6 +35,18 @@ endif()
 add_executable(viam-camera-realsense main.cpp)
 target_link_libraries(viam-camera-realsense PRIVATE viamrealsense)
 
+if(ENABLE_SANITIZER)
+    add_compile_options(-fsanitize=address)
+    add_link_options(-fsanitize=address)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -static-libasan")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -static-libasan")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static-libasan")
+    target_compile_options(viamrealsense PRIVATE -fsanitize=address)
```
Comment (hexbabe) [inline CMakeLists.txt:44]: we can take out this flag if we don't want to compile Intel's lib with ASAN

File: packaging/appimages/viam-camera-realsense-aarch64.yml [hunk 5]
```diff
@@ -42,6 +42,7 @@ AppDir:
   runtime:
     env:
         AIX_TARGET: usr/bin/viam-camera-realsense
+        LSAN_OPTIONS: verbosity=1:log_threads=1
```
Comment (nicksanford) [inline packaging/appimages/viam-camera-realsense-aarch64.yml:45]: The docs say that leak sanitizer works on x86 linux & macos. Dos this do anything on arm?

File: README.md [hunk 6]
```diff
@@ -164,3 +164,6 @@ If you would like to try to gather all of the dependencies yourself and not use
 
 then do `make viam-camera-realsense` to compile the binary, and `make appimage` to create the AppImage.
 
+## Building with Address Sanitizer
+
+When developing, you also have the option to build the module with ASAN/LSAN enabled to help test for memory leaks. You can do so by running `canon -arch arm64 make clean appimage-arm64 SANITIZE=ON`. ASAN/LSAN logs will then be included as error logs in your robot logs.
```
Comment (nicksanford) [inline README.md:169]: Can you please add which platforms ASAN is supported on?

File: CMakeLists.txt (hunk 3 above)
Comment (nicksanford) [reply CMakeLists.txt:47]: You might want to consider adding: 
[-fsanitize-address-use-after-scope](https://github.com/google/sanitizers/wiki/AddressSanitizerUseAfterScope) and [-fno-omit-frame-pointer](https://releases.llvm.org/8.0.0/tools/clang/docs/MemorySanitizer.html)

see: https://github.com/viamrobotics/viam-cartographer/blob/ee4ea28bc5f981d9ddbd93aa7db6a2d1fb95e808/Makefile#L121

File: CMakeLists.txt [hunk 7]
```diff
@@ -32,6 +35,18 @@ endif()
 add_executable(viam-camera-realsense main.cpp)
 target_link_libraries(viam-camera-realsense PRIVATE viamrealsense)
 
+if(ENABLE_SANITIZER)
+    add_compile_options(-fsanitize=address)
+    add_link_options(-fsanitize=address)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -static-libasan")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -static-libasan")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static-libasan")
```
Comment (nicksanford) [inline CMakeLists.txt:43]: Why is `-static-libasan` needed?

File: CMakeLists.txt (hunk 4 above)
Comment (nicksanford) [reply CMakeLists.txt:44]: I think this is fine

File: packaging/appimages/viam-camera-realsense-aarch64.yml (hunk 5 above)
Comment (nicksanford) [reply packaging/appimages/viam-camera-realsense-aarch64.yml:45]: https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizer

Comment (nicksanford) [reply packaging/appimages/viam-camera-realsense-aarch64.yml:45]: Also, why are we only configuring leak sanitizer?

**Review by nicksanford — COMMENTED** [review]
@hexbabe can you please add in the PR description:
1. What manual testing have you done for the asan build?
2. On which CPU / OS platforms?

File: packaging/appimages/viam-camera-realsense-aarch64.yml (hunk 5 above)
Comment (nicksanford) [reply packaging/appimages/viam-camera-realsense-aarch64.yml:45]: As an example, here is how to configure ASAN and some config options I've found useful in the past: https://github.com/viamrobotics/viam-cartographer/blob/ee4ea28bc5f981d9ddbd93aa7db6a2d1fb95e808/Makefile#L122

File: CMakeLists.txt (hunk 7 above)
Comment (hexbabe) [reply CMakeLists.txt:43]: A lot of the current changes are derived from Dan's document [here](https://viam.atlassian.net/wiki/spaces/~71202008791ab3334b43cd8f0797eb00cb641e/pages/330563585/Compiling+Running+a+C+C+module+with+ASAN)

**Comment by nicksanford** [comment]
What C compiler(s) do we support on the intel realsense?

**Comment by nicksanford** [comment]
Can you confirm that there are no new dynamic lib deps added to the module after adding sanitizer?

**Comment by hexbabe** [comment]
> What C compiler(s) do we support on the intel realsense?

From logs: `-- The CXX compiler identification is GNU 12.2.0` So gcc.

The Realsense build process uses canon inside a docker container defined [here](https://github.com/viamrobotics/viam-camera-realsense/tree/main/etc)

**Comment by nicksanford** [comment]
@hexbabe can you please test if the debug build works on all platforms the intel realsense module currently targets & add to the README if the debug build doesn't work on them?

File: packaging/appimages/viam-camera-realsense-aarch64.yml (hunk 5 above)
Comment (hexbabe) [reply packaging/appimages/viam-camera-realsense-aarch64.yml:52]: Configured ASAN as well in the latest commit

File: CMakeLists.txt (hunk 3 above)
Comment (hexbabe) [reply CMakeLists.txt]: Done

**Comment by nicksanford** [comment]
> > @hexbabe can you please confirm that tests FAIL if there is a memory leak introduced on all platforms we support? I don't mean in CI, I mean wherever we support running tests (like on your laptop).
> 
> @nicksanford
> 
> I am currently unable to get this ASAN/LSAN debug build to work on amd64. (For manager/PM mode Nick) Thoughts on how much time I should allocate to trying to get it to work?
> 
> If after the allocated time I am still unable, I can put on the README that the debug build is currently only verified to work on linux/arm64

If you are hitting friction, then don't worry about amd64 support for now, just document that the ASAN build doesn't work on amd64.

File: .gitignore [hunk 8]
```diff
@@ -3,4 +3,5 @@ module.tar.gz
 integration/realsense-integration-tests
 realsense-integration-tests
 build
+local-build
```
Comment (nicksanford) [inline .gitignore:6]: why is this being added?

Comment (hexbabe) [reply .gitignore]: My bad that was for my local testing to organize my source control. Removed.

**Comment by nicksanford** [comment]
@hexbabe please remove:
viam-camera-realsense and packaging/appimages/viam-camera-realsense-latest-x86_64.AppImage.zsync

**Comment by hexbabe** [comment]
> @hexbabe please remove: viam-camera-realsense and packaging/appimages/viam-camera-realsense-latest-x86_64.AppImage.zsync

Removed all artifacts

File: README.md (hunk 6 above)
Comment (hexbabe) [reply README.md]: Done

**Review by nicksanford — APPROVED** [review]

**Pull request merged**